### ISSUE TITLE
Update vulnerable dependencies (or use alternatives)

### ISSUE THREAD

**Issue by ChALkeR** [firebase/firebase-tools#769]
See https://snyk.io/test/npm/firebase-tools — it lists 4 vulnerable deps (2 high, 2 low) installed with the latest version of `firebase-tools`.

Related: #235, but that lists other deps (probably no longer valid?).

Current Snyk output:
1. [High] Arbitrary Command Injection
    Introduced through: `firebase-tools@3.18.4 › open@0.0.5`,
1. [High] Directory Traversal
    Introduced through: `firebase-tools@3.18.4 › superstatic@5.0.1`,
1. [Low] Prototype Pollution
    Introduced through: `firebase-tools@3.18.4 › jsonwebtoken@7.4.3 › joi@6.10.1 › hoek@2.16.3`,
1. [Low] Prototype Pollution
    Introduced through: `firebase-tools@3.18.4 › superstatic@5.0.1 › nash@2.0.4 › lodash@3.10.1` and 2 other lodash paths.

**Comment by ChALkeR** [comment]
This is what `npm install` prints now:

```console
$ npm i firebase-tools

…
…

+ firebase-tools@3.18.4
added 533 packages from 270 contributors in 53.092s
[!] 5 vulnerabilities found [2483 packages audited]
    Severity: 2 low | 2 moderate | 1 high
    Run `npm audit` for more detail
```

![screenshot_20180516_204017](https://user-images.githubusercontent.com/291301/40133764-58d2125e-5949-11e8-92d0-21b261541eee.png)

`npm audit` output:
```console
$ npm audit
                                                                                
                       === npm audit security report ===                        
                                                                                



┌──────────────────────────────────────────────────────────────────────────────┐
│                                Manual Review                                 │
│            Some vulnerabilities require your attention to resolve            │
│                                                                              │
│         Visit https://go.npm.me/audit-guide for additional guidance          │
└──────────────────────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Prototype pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ hoek                                                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ firebase-tools                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ firebase-tools > jsonwebtoken > joi > hoek                   │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/566                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Prototype pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ hoek                                                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ firebase-tools                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ firebase-tools > jsonwebtoken > joi > topo > hoek            │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/566                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ high          │ Path Traversal                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ superstatic                                                  │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ firebase-tools                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ firebase-tools > superstatic                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/652                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ low           │ Prototype Pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ lodash                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ firebase-tools                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ firebase-tools > superstatic > glob-slasher > toxic > lodash │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/577                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ low           │ Prototype Pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ lodash                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ firebase-tools                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ firebase-tools > superstatic > nash > lodash                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/577                       │
└───────────────┴──────────────────────────────────────────────────────────────┘

[!] 5 vulnerabilities found - Packages audited: 2483 (0 dev, 1275 optional)
    Severity: 2 low | 2 moderate | 1 high
```

**Comment by laurenzlong** [comment]
Unfortunately superstatic and open are already at their latest versions, so this is not as simple as updating dependencies. I don't  think we can get away from using superstatic, but we may be able to use an alternative for open, PRs welcome :)

**Comment by ChALkeR** [comment]
@laurenzlong `superstatic` is a part of `firebase`, so that is in scope — just release a new version.

`open` could be replaced.

I am not a user of `firebase-tools` and do not care enough to file a PRs, sorry.
Especially given the lack of interest from maintainers.

**Comment by merlinnot** [comment]
> Especially given the lack of interest from maintainers.

That was harsh. It's probably not their fault, Google has other priorities... They release some half-baked stuff and then they move employees to different projects. You can expect a high quality from Google. Take a look a Polymerfire, it's in much, much worse condition ;)

**Comment by ChALkeR** [comment]
@merlinnot Why «harsh»? I believe that I'm being objective here.

I reported a vuln, which they have a landed patch for and that they disclosed 115 days ago. They didn't release a fixed version yet to consumers, it's stuck on https://github.com/firebase/superstatic/pull/255.
I tried pinging them multiple times about it, even got some responses, but those did not lead to any visible progress..

**Comment by mbleigh** [comment]
3.19.1 has been released and addresses vulnerable deps.

**Comment by ChALkeR** [comment]
@mbleigh I think there is one left (other than superstatic path traversal that should have the data updated):
```
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ High          │ Path Traversal                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ superstatic                                                  │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ No patch available                                           │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ firebase-tools                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ firebase-tools > superstatic                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/652                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Low           │ Prototype Pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ lodash                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=4.17.5                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ firebase-tools                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ firebase-tools > superstatic > nash > lodash                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/577                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
```

**Comment by mbleigh** [comment]
Yes, we still have that one to deal with, but it doesn't directly affect `firebase-tools` (it is used by a package that is only used to control the `superstatic` command-line interface but is never loaded in the course of using `firebase-tools`) and is marked as "Low".

Will try to patch it through Superstatic but wanted to get a fix out for the higher level vulnerabilities ASAP.

**Comment by cjcurrie** [comment]
I have four Prototype Pollution high-level warnings from node-forge via firebase-tools
They are all very similar to 
```
  High            Prototype Pollution in node-forge

  Package         node-forge

  Patched in      >= 0.10.0

  Dependency of   firebase-tools

  Path            firebase-tools > google-gax > google-auth-library > gtoken >
                  google-p12-pem > node-forge

  More info       https://npmjs.com/advisories/1561
```
